File: 0817-linked-list-components/0817-linked-list-components.cpp
#include <unordered_set>

// Definition for singly-linked list.
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(NULL) {}
// };

class Solution {
public:
    int numComponents(ListNode* head, vector<int>& nums) {
        // Convert nums to an unordered_set for O(1) look-up
        std::unordered_set<int> num_set(nums.begin(), nums.end());
        
        // Initialize variables
        int count = 0;
        bool connected = false;
        
        // Traverse the linked list
        while (head) {
            // If the current value is in nums
            if (num_set.find(head->val) != num_set.end()) {
                // If not already connected, increment the count
                if (!connected) {
                    ++count;
                }
                // Mark as connected
                connected = true;
            } else {
                // Mark as not connected
                connected = false;
            }
            
            // Move to the next node
            head = head->next;
        }
        
        return count;
    }
};

File: 0817-linked-list-components/0817-linked-list-components.py
from typing import List, Optional

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def numComponents(self, head: Optional[ListNode], nums: List[int]) -> int:
        # Convert nums to a set for O(1) look-up
        num_set = set(nums)
        
        # Initialize variables
        count = 0
        connected = False
        
        # Traverse the linked list
        while head:
            # If the current value is in nums
            if head.val in num_set:
                # If not already connected, increment the count
                if not connected:
                    count += 1
                # Mark as connected
                connected = True
            else:
                # Mark as not connected
                connected = False
            
            # Move to the next node
            head = head.next
        
        return count
